### PR TITLE
Time - Emily

### PR DESCRIPTION
**Pull request by snowistaken**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | One custom model method that I wrote was for finding the spotlight work. This method is in the work model.
Describe how you approached testing that model method. What edge cases did you come up with? | Some edge cases that I came up with were ensuring that a work was still returned if there were no voted works, as well as ensuring that the first work is selected if there is a tie.
What are `session` and `flash`? What is the difference between them? | Session is a hash like object that persists through the user's visit to the site. This information is stored in the cookies in their browser. This allows us to know whether or not they are logged in, as well as to log them in when they attempt to do so. Flash is a tool that allows us to notify the user of something when they make a request.
What was one thing that you gained more clarity on through this assignment? | Relationships and model methods are two things that I gained clarity on in this project. Understanding where to put methods and which ones to have makes the overall workflow of the project a lot more efficient.
What is the Heroku URL of your deployed application? | I forgot to deploy my most recent version! Will update this with link at the same time that I fix tests.

My model tests are not currently complete. I'll be updating my PR with the completed tests this evening. Thanks :)

### CODE REVIEW

File: test/models/vote_test.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+require "test_helper"
+
+describe Vote do
+  # it "does a thing" do
+  #   value(1+1).must_equal 2
+  # end
+
+  describe 'validations' do
+    before do
+      @vote = Vote.new(user_id: 1, work_id: 1)
```
Comment (CheezItMan) [inline test/models/vote_test.rb:10]: This won't work because you can't guarantee that the test database will have a user and work with id 1.

Either:

1. Create fixtures for your test database and use them or
1. Do something like this

```suggestion
      @user = User.create(username: 'bob');
      @user = User.create(username: 'bob');
      @work = Work.create(category: 'book', title: 'Wizard of Oz', creator: 'Someone', publication_year: "1920", description: 'a book about a girl who got hit on the head')
      @vote = Vote.new(user_id: @user.id, work_id: @work.id)
```

File: test/models/work_test.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      @work = Work.new(category: "test category", title: "test work", creator: "author", publication_year: "01-01-2000", description: "test")
+    end
+
+    it 'is valid when all fields are present' do
+      result = @work.valid?
+
+      expect(result).must_equal true
+    end
+
+    it 'is invalid without a category' do
```
Comment (CheezItMan) [inline test/models/work_test.rb:19]: Just a note, that you can loop through the required fields, you don't have to make a test for each one. 

```ruby
  required_fields = [:category, :title, :creator, :publication_year]
  required_fields.each do |field|
    old_value = @work[field]
    @work[field] = nil
    expect(@work.valid?).must_equal false
    @work[field] = old_value
  end

```

File: test/models/work_test.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    it 'remains valid without a description' do
+      @work.description = nil
+
+      result = @work.valid?
+
+      expect(result).must_equal true
+    end
+  end
+
+  describe 'work model methods' do
```
Comment (CheezItMan) [inline test/models/work_test.rb:64]: Just noting these aren't done yet.

File: test/models/user_test.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      @user.username = nil
+
+      result = @user.valid?
+
+      expect(result).must_equal false
+      expect(@user.errors.messages).must_include :username
+    end
+  end
+
+  # describe 'relations' do
```
Comment (CheezItMan) [inline test/models/user_test.rb:30]: Just noting these are missing.

File: app/models/vote.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class Vote < ApplicationRecord
+  validates :user_id, presence: true
+  validates :work_id, presence: true
```
Comment (CheezItMan) [inline app/models/vote.rb:3]: You don't need to have these, the relationships require them anyway.

However you **do** need a validation to prevent a duplicate combination user_id and work_id.

```suggestion
   #  Requires every combination work_id and user_id to be unique.
  validates :user_id, uniqueness: {scope: :work_id}
```

File: app/models/work.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
+class Work < ApplicationRecord
+  validates :title, presence: true, uniqueness: true
+  validates :category, presence: true
+  validates :publication_year, presence: true
+  validates :creator, presence: true
+  has_many :votes
+
+  def self.spotlight
+    all_works = Work.all
+    result = all_works.sort_by { |work| work.votes.length }.reverse
```
Comment (CheezItMan) [inline app/models/work.rb:10]: This works and is fine right now, but it's better if you use the database methods like `.order` and `where` etc to handle these operations.  The database is optimized for this much better.

File: app/views/works/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  <h2>List of Works</h2>
+
+  <div>
+    <h3>Albums</h3>
+    <ul>
+      <% @albums.each do |album| %>
+          <li>
+            <%= link_to album.title, work_path(album) %> 
+          </li>
+        <% end %>
```
Comment (CheezItMan) [inline app/views/works/index.html.erb:11]: Notice that this stuff is pretty repetitive, so you could actually build a view helper for it or a partial view.

File: config/routes.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+Rails.application.routes.draw do
+  root to: "homepage#index"
+  resources :homepage, only: [:index]
```
Comment (CheezItMan) [inline config/routes.rb:3]: If you  have the root route going to the homepage you don't need this.

```suggestion
```

File: config/routes.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+  resources :works
+
+  post "/upvote/:id", to: "works#upvote", as: "upvote"
```
Comment (CheezItMan) [inline config/routes.rb:7]:
```suggestion
  post "works/upvote/:id", to: "works#upvote", as: "upvote"
```

File: app/controllers/works_controller.rb [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    @books = Work.sort_by_category("book")
+    @movies = Work.sort_by_category("movie")
+    @albums = Work.sort_by_category("album")
+  end
+
+  def show
+    @work = Work.find_by(id: params[:id])
+
+    if @work.nil?
+      head :not_found
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:12]: In the future I advise just redirecting the user rather than just presenting them with a blank white page.

File: app/controllers/works_controller.rb [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+      return
+    elsif @work.update(work_params)
+      redirect_to work_path(@work.id)
+      return
+    else
+      render :edit
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:36]: It's best to include a response code for all renders, and also use `flash.now` to present validation errors to the user.

```suggestion
      flash.now[:error] = @work.errors.messages
      render :edit, status: :bad_request
```

File: app/controllers/works_controller.rb [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  end
+
+  def create
+    @work = Work.new(work_params)
+
+    if @work.save
+      redirect_to work_path(@work.id)
+      return
+    else
+      render :new
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:64]:
```suggestion
      render :new, status: :bad_request
```

File: test/controllers/homepage_controller_test.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe HomepageController do
+  # it "does a thing" do
+  #   value(1+1).must_equal 2
+  # end
```
Comment (CheezItMan) [inline test/controllers/homepage_controller_test.rb:6]: Just noting this controller is untested.

File: test/controllers/users_controller_test.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe UsersController do
+  # it "does a thing" do
+  #   value(1+1).must_equal 2
+  # end
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb:6]: Just noting no tests in the users controller

File: app/controllers/works_controller.rb [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    end
+  end
+
+  def upvote
+    @current_user = User.find_by(id: session[:user_id])
+    @work = Work.find_by(id: params[:id])
+
+    if @current_user.nil?
+      # redirect_to work_path(@work.id)
+      flash[:error] = "You must be logged in to do that!"
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:75]: After you use `flash` you should do a redirect.

```suggestion
      flash[:error] = "You must be logged in to do that!"
      redirect_back fallback_location: works_path
```

File: app/controllers/works_controller.rb [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    elsif Work.upvoted?(@current_user.id, @work.id)
+      flash[:error] = "You have already voted for this work!"
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:77]:
```suggestion
      flash[:error] = "You must be logged in to do that!"
      redirect_back fallback_location: works_path
```

File: app/models/work.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    return result.slice(0, 10)
+  end
+
+  def self.upvoted?(user_id, work_id)
+    if Vote.find_by(user_id: user_id, work_id: work_id)
+      return true
+    else
+      return false
+    end
+  end
```
Comment (CheezItMan) [inline app/models/work.rb:27]: This is better to do with a validation instead of a method.

File: app/controllers/works_controller.rb [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+      head :not_found
+      return
+    end
+  end
+
+  def update
+    @work = Work.find_by(id: params[:id])
+
+    if @work.nil?
+      head :not_found
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:30]:
```suggestion
      flash[:notice] = "Work not found"
      redirect_to root_path
```

File: app/controllers/works_controller.rb [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+      return
+    elsif @work.update(work_params)
+      redirect_to work_path(@work.id)
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:33]: I suggest adding a flash notice here.

File: app/controllers/works_controller.rb [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+      return
+    else
+      render :edit
+      return
+    end
+  end
+
+  def destroy
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:41]: Flash notices would be good here.

File: app/views/layouts/application.html.erb [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    <li class="nav-item">
+      <a>Logged in as: <%= User.find_by(id: session[:user_id]).username %> |</a>
+    </li>
+  <% end %>
+  <% if session[:user_id] %>
+    <li class="nav-item">
+      <a><%= link_to "Log Out", logout_path, method: :post%></a>
+    </li>
+  <% end %>
+</ul>
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:43]: This content should be in the `body` tag to be proper HTML.

File: app/views/works/index.html.erb [hunk 22]
```diff
@@ -0,0 +1,36 @@
+<body>
```
Comment (CheezItMan) [inline app/views/works/index.html.erb:1]: The body tag is in the layout file and there should only be one body in the HTML document. 

Perhaps you meant `main`

**Comment by CheezItMan** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ⚠, This doesn't work see my notes, you forgot to redirect after setting the flash notice. 
| 4. Can edit this work successfully, and get a flash message | ⚠, no flash notice presented
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ⚠, no validation errors presented.  
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ⚠, No flash notice here
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ⚠ changed the UI, but didn't show a flash notice
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ⚠ no flash notice provided

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ⚠ no Deployment at present
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ⚠ The styling is minimal.  However the functionality is **mostly** the same. 
| 3. Practices git with at least 25 small commits and meaningful commit messages | ⚠ Only a few commits.  For a project like MediaRanker, I would expect at least 30 commits

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ⚠ You're using a lot of `div` tags and improperly using the `body` tag.  
| 7. The homepage view, all media view, and new works view use partials when appropriate | ⚠ you're only using partials in forms, there are a few other places to use one. 
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ⚠ Not there
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ⚠ Not there
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️, See my notes on flash.

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ⚠ No fixtures
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ⚠ Missing

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | I judged this as yellow based on some parts only partially completed.






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Descriptive/Readable | ✅
| Logical/Organized | ✅


## Summary


You got the basic functionality completed Emily.  See my comments in your code below.  From this I think you need more practice with testing including fixtures, custom methods and controller tests.  You also need more practice writing flash notices and styling.  That said you have the relationships working and you have some custom validations and business logic methods.  I'm glad you got this in and if you push any further commits and want my feedback just ping me in slack.
